bevy_lint/src/lints/insert_event_resource.rs

```diff
@@ -0,0 +1,226 @@
+//! #
+//! #[derive(Event)]
+//! struct MyEvent;
+//!
+//! App::new().add_event::<MyEvent>().run();
+//! ```
+
+use clippy_utils::{
+    diagnostics::span_lint_and_sugg, source::snippet_with_applicability, sym, ty::match_type,
+};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, GenericArg, GenericArgs, Path, PathSegment, QPath};
+use rustc_hir_analysis::lower_ty;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{Ty, TyKind};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::Span;
+use std::borrow::Cow;
+
+declare_tool_lint! {
+    pub bevy::INSERT_EVENT_RESOURCE,
+    Deny,
+    "called `App::insert_resource(Events<T>)` or `App::init_resource::<Events<T>>()` instead of `App::add_event::<T>()`"
+}
+
+declare_lint_pass! {
+    InsertEventResource => [INSERT_EVENT_RESOURCE]
+}
+
+impl<'tcx> LateLintPass<'tcx> for InsertEventResource {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &Expr<'tcx>) {
+        // Find a method call.
+        if let ExprKind::MethodCall(path, src, args, method_span) = expr.kind {
+            // Get the type for `src` in `src.method()`. We peel all references because the type
+            // could either be `App` or `&mut App`.
+            let src_ty = cx.typeck_results().expr_ty(src).peel_refs();
+
+            // If `src` is not a Bevy `App`, exit.
+            if !match_type(cx, src_ty, &crate::paths::APP) {
+                return;
+            }
+
+            // If the method is `App::insert_resource()` or `App::init_resource()`, check it with
+            // its corresponding function.
+            match path.ident.name {
+                symbol if symbol == sym!(insert_resource) => {
+                    check_insert_resource(cx, args, method_span)
+                }
+                symbol if symbol == sym!(init_resource) => {
+                    check_init_resource(cx, path, method_span)
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+/// Checks if `App::insert_resource()` inserts an `Events<T>`, and emits a diagnostic if so.
+fn check_insert_resource<'tcx>(cx: &LateContext<'tcx>, args: &[Expr], method_span: Span) {
+    // Extract the argument if there is only 1 (which there should be!), else exit.
+    let [arg] = args else {
+        return;
+    };
+
+    // Find the type of `arg` in `App::insert_resource(arg)`.
+    let ty = cx.typeck_results().expr_ty(arg);
+
+    // If `arg` is `Events<T>`, emit the lint.
+    if match_type(cx, ty, &crate::paths::EVENTS) {
+        let mut applicability = Applicability::MachineApplicable;
+
+        let event_ty_snippet = extract_ty_event_snippet(ty, &mut applicability);
+
+        span_lint_and_sugg(
+            cx,
+            INSERT_EVENT_RESOURCE,
+            method_span,
+            "called `App::insert_resource(Events<T>)` instead of `App::add_event::<T>()`",
+            "inserting an `Events` resource does not fully setup that event",
```

Comment:
```suggestion
            "`add_event` will initialize the `Events` `Resource` as well as other necessary behavior"
```

Reply:
Thoughts on "`App::add_event()` will initialize `Events<T>` and prevent memory leaks" instead, as it's more specific?

Reply:
Hmm, this is missing the feeling of "the offending code will set stuff up in an incomplete manner". As a reader, this sounds to me like "Okay, why would I care about that `Resource`? "necessary behavior" for what? Using the events at all, or some advanced features?" The shorter version is a bit more absolute in this regard: "You may think you've done a setup, but you actually didn't".
Again, this is not a strong opinion, feel free to disagree.

Reply:
^ That message was directed at the original suggestion. @BD103 I like your suggestion better, but it seems a bit... technical? Like, isn't it enough to say that the code in question will initialize things incompletely and then have the details of what that exactly means in the documentation?

Reply:
Based off of Jan's comment, I'm going to leave it as-is. It's less wordy, and there's plenty of other documentation that will resolve any confusion.